## Content

[01-raw_Illumina_QC](https://github.com/EmelineFavreau/Pheidole_pallidula_social_polymorphism/tree/master/01-raw_Illumina_QC): Quality control of short reads for population genetics analyses   
[02-species_identification](https://github.com/EmelineFavreau/Pheidole_pallidula_social_polymorphism/tree/master/02-species_identification): Using microsatellites for taxonomic identification  
[03-raw_MinION_basecalling](https://github.com/EmelineFavreau/Pheidole_pallidula_social_polymorphism/tree/master/03-raw_MinION_basecalling): For genome assembly  
[04-genome_assembly](https://github.com/EmelineFavreau/Pheidole_pallidula_social_polymorphism/tree/master/04-genome_assembly): Assembling, polishing, scaffolding, measuring N50  
[05-population_genetics_analysis](https://github.com/EmelineFavreau/Pheidole_pallidula_social_polymorphism/tree/master/05-population_genetics_analysis):  Including FST, Fisher's exact tests and simulation models  
[06-checking_absence_of_supergene](https://github.com/EmelineFavreau/Pheidole_pallidula_social_polymorphism/tree/master/06-checking_absence_of_supergene):  Including measuring read coverage variation

## Context & citation

This repository contains all scripts used for the population genomic analysis of *Pheidole pallidula* sampled across Europe. This was developed as part of a NERC-funded PhD project implemented in Yannick Wurm's lab at Queen Mary University of London. Findings are in a submitted manuscript.


