---
title: "Pheidole microsatellite PCA"
author: "Emeline Favreau"
date: "01/07/2019"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r import all the data , eval = TRUE, echo = FALSE, include = FALSE}
# import coverage for monogynous - this takes few minutes
all_colonies_microsatellites <- read.csv("../data/all_colonies_genotype.csv")


# load all the libraries
# get libraries adegenet 2.1.1
basic_libraries <- c("adegenet", "ggplot2", "wordcloud")
for (lib in basic_libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                install.packages(lib)
                library(lib, character.only = TRUE )
        }
}
```

We genotyped 115 colonies for number of queens. We want to investigate if there is any outlier in our data.


```{r test, eval = TRUE, echo = FALSE, warning = FALSE, width = 20, height = 5}
# obtain name of loci pairs
loci_pair_vec <- grep(x = colnames(all_colonies_microsatellites), pattern = "p_", value = TRUE)

# obtain name of loci
loci_vec <- unique(gsub(x = loci_pair_vec, pattern = "_0(1|2)$", replacement = ""))

# obtain info for one colony
#one_colony_microsatellites <- subset(all_colonies_microsatellites, subset = col_name == "cmr00001", select = c("w_name", loci_pair_vec))
one_colony_microsatellites <- subset(all_colonies_microsatellites, select = c("w_name", loci_pair_vec))


# format alleles to be imported by adegenet package

# obtain loop parameters (select two columns at once)
loop_parameters <- seq(1, length(loci_pair_vec), by = 2)

# set a vector to be filled with number of alleles per locus
one_colony_microsatellites_adagenet_mat <- c()

# number of alleles per primer
for(this_primer in loop_parameters){
  # subset dataframe for just this locus
  one_colony_microsatellites_one_locus <- subset(one_colony_microsatellites, select = colnames(one_colony_microsatellites) %in% loci_pair_vec[c(this_primer, this_primer+1)])
  # format alleles to adegenet standard
  allele_reformated <- paste(one_colony_microsatellites_one_locus[, 1], one_colony_microsatellites_one_locus[, 2], sep = "/")
  # add it to the vector
  one_colony_microsatellites_adagenet_mat <- cbind(one_colony_microsatellites_adagenet_mat, allele_reformated)
}


# set dimnames for this matrix
dimnames(one_colony_microsatellites_adagenet_mat) <- list(one_colony_microsatellites$w_name, loci_vec)

# make a genind object
obj <- df2genind(one_colony_microsatellites_adagenet_mat, ploidy = 2, sep = "/")

pop(obj) <- all_colonies_microsatellites$Locality
summary(obj)
# our object into a genpop
obj2 <- genind2genpop(obj)

# Allele presence absence data are extracted and NAs replaced using tab
X <- tab(obj, NA.method = "mean")

## make PCA on allele presence/absence
# principal component analysis of a data frame and returns the results as objects of class pca and dudi.
pca1 <- dudi.pca(X, scannf = FALSE, scale = FALSE)

# Graph of individuals. Individuals with a similar profile are grouped together.
# create data frame with scores
scores <- as.data.frame(pca1$li)

# add rownames
rownames(scores)  <- paste(obj@pop, 1:nrow(scores), sep = "")

# add population
scores$population <- obj@pop

# add colony name
scores$colony_name <- all_colonies_microsatellites$col_name

# add gyny
scores$gyny <- all_colonies_microsatellites$gyny

# name columns
colnames(scores) <- c("PC1", "PC2", "population", "colony_name", "gyny")

# eight colours that are diverging and color-bling
eight_colours <- c('#8c510a','#bf812d','#dfc27d','#f6e8c3','#c7eae5','#80cdc1','#35978f','#01665e')

# plot of observations
ggplot(data = scores, aes(x = PC1, y = PC2, label = rownames(scores), colour = population, shape = gyny)) +
  geom_point() +
  ggtitle("PCA from Pheidole microsatellites") +
  theme_classic() +
  scale_color_manual(values = eight_colours)




```
