---
title: "Association analysis Pheidole + fake SNPs"
author: "Emeline Favreau"
date: "8th October 2020"
output: pdf_document
  
---

I tested each SNP for association with social type using Fisher exact test implemented in PLINK.
124,840 biallelic SNPs, supported by at least 75% of samples, socially polymorphic within each geographical population. 
3,054 of the SNPs are simulated, based on _Solenopsis invicta_'s system.
Here I test whether the simulated SNPs are significantly associated with social type.


```{r load libraries, eval = TRUE, echo = FALSE, include = FALSE}
# check the working directory
basic_libraries <- c("qqman",
                     "ggplot2",
                     "tidyverse")

for (lib in basic_libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                install.packages(lib)
                library(lib, character.only = TRUE )
        }
}
```


```{r create my own manhattan function, eval = TRUE, echo = FALSE}
# create my own manhattan function
em_manhattan <- function(x, chr = "CHR", bp = "BP", p = "P", snp = "SNP", col = c("gray10", 
    "gray60"), chrlabs = NULL, suggestiveline = -log10(1e-05), 
    genomewideline = -log10(5e-08), highlight = NULL, logp = TRUE, 
    annotatePval = NULL, annotateTop = TRUE, pointshape = 20, sigshape = 15, ...) 
{
    # checking the inputs are fine  
    CHR = BP = P = index = NULL
    if (!(chr %in% names(x))) 
        stop(paste("Column", chr, "not found!"))
    if (!(bp %in% names(x))) 
        stop(paste("Column", bp, "not found!"))
    if (!(p %in% names(x))) 
        stop(paste("Column", p, "not found!"))
    if (!(snp %in% names(x))) 
        warning(paste("No SNP column found. OK unless you're trying to highlight."))
    if (!is.numeric(x[[chr]])) 
        stop(paste(chr, "column should be numeric. Do you have 'X', 'Y', 'MT', etc? If so change to numbers and try again."))
    if (!is.numeric(x[[bp]])) 
        stop(paste(bp, "column should be numeric."))
    if (!is.numeric(x[[p]])) 
        stop(paste(p, "column should be numeric."))
    # create a dataframe with plotting info
    d = data.frame(CHR = x[[chr]], BP = x[[bp]], P = x[[p]])
    if (!is.null(x[[snp]])) 
        d = transform(d, SNP = x[[snp]])
    d <- subset(d, (is.numeric(CHR) & is.numeric(BP) & is.numeric(P)))
    # order first by chromosome then by base pair
    d <- d[order(d$CHR, d$BP), ]
    # transform p value in log if command present
    if (logp) {
        d$logp <- -log10(d$P)
    }
    else {
        d$logp <- d$P
    }
    d$pos = NA
    d$index = NA
    ind = 0
    # add a individual rank for each SNP, following the order of CHR and BP
    for (i in unique(d$CHR)) {
        ind = ind + 1
        d[d$CHR == i, ]$index = ind
    }
    nchr = length(unique(d$CHR))
    # if there is only 1 chromosome
    if (nchr == 1) {
        d$pos = d$BP
        ticks = floor(length(d$pos))/2 + 1
        xlabel = paste("Chromosome", unique(d$CHR), "position")
        labs = ticks
    }
    # if there are more than 1 chromosome
    else {
        lastbase = 0
        ticks = NULL
        for (i in unique(d$index)) {
            if (i == 1) {
                d[d$index == i, ]$pos = d[d$index == i, ]$BP
            }
            else {
                # set a value for the last base
                lastbase <- lastbase + tail(subset(d, index == 
                  i - 1)$BP, 1)
                # set a value for position
                d[d$index == i, ]$pos <- d[d$index == i, ]$BP + 
                  lastbase
            }
          # give specific values to the ticks in x axis
            ticks = c(ticks, (min(d[d$index == i, ]$pos) + max(d[d$index == 
                i, ]$pos))/2 + 1)
        }
        xlabel = "Chromosome"
        labs <- unique(d$CHR)
    }
    # set the x axis limits
    xmax = ceiling(max(d$pos) * 1.03)
    xmin = floor(max(d$pos) * -0.03)
    # set the default arguments
    def_args <- list(xaxt = "n",
                     #yaxt = "n",
                     # no box around the plot
                     bty = "n",
                     xaxs = "i",
                     yaxs = "i", 
                     # labels are always horizontal
                     las = 1,
                     # point shape: filled, round dot
                     pch = pointshape,
                     # axis limits
                     xlim = c(xmin, xmax),
                     ylim = c(0, ceiling(max(d$logp))),
                     xlab = xlabel,
                     ylab = "") 
    # create a list with the extra arguments listed by user
    dotargs <- list(...)
    # plot
    do.call("plot", c(NA, dotargs, def_args[!names(def_args) %in% 
        names(dotargs)]))
    
    if (!is.null(chrlabs)) {
      # give warnings if things are off
        if (is.character(chrlabs)) {
            if (length(chrlabs) == length(labs)) {
                labs <- chrlabs
            }
            else {
                warning("You're trying to specify chromosome labels but the number of labels != number of chromosomes.")
            }
        }
        else {
            warning("If you're trying to specify chromosome labels, chrlabs must be a character vector")
        }
    }
    # add specific x axis
    if (nchr == 1) {
        axis(1, ...)
    }
    # specify the ticks here for the x axis (side = 1)
    else {
        axis(1, at = ticks, labels = labs, tck = -0.025, cex.axis = 1)
      
    }
    # add text for p-value threshold
    #text(75, suggestiveline + 0.1, "p-value threshold = 0.05",
    # cex = .8, pos = 4, col = "purple")
    
    # specify the change of colours at each chromosome
    col = rep(col, max(d$CHR))
    # if there is only 1 chromosome, there will by only one colour
    if (nchr == 1) {
        with(d, points(pos, logp, pch = pointshape, col = col[1], ...))
    }
    # if there are more than 1 chrosomome, there will be a change of colour
    else {
        icol = 1
        for (i in unique(d$index)) {
            with(d[d$index == unique(d$index)[i], ], points(pos, 
                logp, col = col[icol], pch = pointshape, ...))
            icol = icol + 1
        }
    }
    
    
    # add a suggestive line
    if (suggestiveline) 
        abline(h = suggestiveline, col = "#018571")
    # add a genomewide line
    if (genomewideline) 
        abline(h = genomewideline, col = "red")
    # colour specific points
    if (!is.null(highlight)) {
        if (any(!(highlight %in% d$SNP))) 
            warning("You're trying to highlight SNPs that don't exist in your results.")
        d.highlight <- d[which(d$SNP %in% highlight), ]
        with(d.highlight, points(pos, logp, col = "blue", pch = sigshape, 
            ...))
    }
    # annotate some points
    if (!is.null(annotatePval)) {
        topHits = subset(d, P <= annotatePval)
        par(xpd = TRUE)
        if (annotateTop == FALSE) {
            with(subset(d, P <= annotatePval), textxy(pos, -log10(P), 
                offset = 0.625, labs = topHits$SNP, cex = 0.45), 
                ...)
        }
        else {
            topHits <- topHits[order(topHits$P), ]
            topSNPs <- NULL
            for (i in unique(topHits$CHR)) {
                chrSNPs <- topHits[topHits$CHR == i, ]
                topSNPs <- rbind(topSNPs, chrSNPs[1, ])
            }
            textxy(topSNPs$pos, -log10(topSNPs$P), offset = 0.5, 
                labs = topSNPs$SNP, cex = 0.5, ...)
        }
    }
    # plot y axis
    title(ylab = expression(-log[10](italic(p))),
          line = 2)
    # This value specifies where in the plotting device an object can actually
    # be plotted. The default is xpd = FALSE, which means that plotting is clipped,
    # or restricted, to the plotting region.
    par(xpd = FALSE)
}

```

## Solenopsis simulation

```{r solenopsis fisher analysis, eval = TRUE, echo = FALSE}
# Evaluating raw p-value distribution from regression analysis
plink_output <- read.csv("result/2020-10-08-fake-Solenopsis-snps-Fisher-Bonferroni.assoc.fisher",
                         header = TRUE,
                         sep = "")
# head(plink_output)

# number of SNPs (check with bcftools view -H 124840)
# nrow(plink_output)

# Bonferroni adjustment
plink_output$P_adj <- p.adjust(p = plink_output$P, method = "bonferroni")

# number of significant SNPs = 3099
nrow(plink_output[plink_output$P_adj < 0.05, ])

# save the significant SNPs name
sig_snp_CHR_names <- plink_output$CHR[plink_output$P_adj < 0.05]

# save the simulated SNPs names
simulated_snp_names <- plink_output$SNP[grep(pattern = "*fake*",
                            x = plink_output$CHR)]

# are all simulated SNPs significant? (ie p adjust < 0.05) 
# expect: 3054 
#length(grep(pattern = "*fake*", x = sig_snp_CHR_names, value = TRUE))

# create one column for numeric value of contigs
# necessary for simple Manhattan plot
plink_output$CHR_num <- as.numeric(plink_output$CHR)

# Manhattan plot
man_title <- paste("Solenopsis simulations with Pheidole SNPs")

man_xlab  <- paste("Contigs")



em_manhattan(plink_output, 
          chr            = "CHR_num",
          bp             = "BP", 
          p              = "P_adj", 
          snp            = "SNP", 
          suggestiveline = -log10(0.05),
          genomewideline = FALSE,
          highlight      = simulated_snp_names,
          main           = man_title, 
          xlab           = man_xlab)



```

### Conclusion
All the SNPs that were simulated following _Solenopsis_ system are significantly associated with social type.
This is what we expected.

## Formica simulation

```{r formica fisher analysis, eval = TRUE, echo = FALSE}
# Evaluating raw p-value distribution from regression analysis
for_plink_output <- read.csv("result/2020-10-12-fake-Formica-snps-Fisher-Bonferroni.assoc.fisher",
                         header = TRUE,
                         sep = "")
# head(for_plink_output)

# number of SNPs (check with bcftools view -H 124840)
# nrow(for_plink_output)

# Bonferroni adjustment
for_plink_output$P_adj <- p.adjust(p = for_plink_output$P,
                                   method = "bonferroni")

# number of significant SNPs = 4505
# nrow(for_plink_output[for_plink_output$P_adj < 0.05, ])

# save the significant SNPs name
for_sig_snp_CHR_names <- for_plink_output$CHR[for_plink_output$P_adj < 0.05]

# save the simulated SNPs names
for_simulated_snp_names <- for_plink_output$SNP[grep(pattern = "*fake*",
                            x = for_plink_output$CHR)]

# are all simulated SNPs significant? (ie p adjust < 0.05) 
# expect: 4505 
#length(grep(pattern = "*fake*", x = for_sig_snp_CHR_names, value = TRUE))

# create one column for numeric value of contigs
# necessary for simple Manhattan plot
for_plink_output$CHR_num <- as.numeric(for_plink_output$CHR)

# Manhattan plot
man_title <- paste("Formica simulations with Pheidole SNPs")

man_xlab  <- paste("Contigs")

em_manhattan(for_plink_output, 
          chr            = "CHR_num",
          bp             = "BP", 
          p              = "P_adj", 
          snp            = "SNP", 
          suggestiveline = -log10(0.05),
          genomewideline = FALSE,
          highlight      = for_simulated_snp_names,
          main           = man_title, 
          xlab           = man_xlab)



```


## Misgenotyping

Misgenotyping could have happen because of null alleles during microsatellite PCR steps. We tested the power of our analysis (Fisher's exact test) for misgenotyping the social type (multiple-queen or single-queen).

We simulated 1000 times our association analysis, for each time 10% of the samples were assigned the alternative social type (e.g. single-queen sample becomes multiple-queen sample).

If our analysis stands misgenotyping, the number of significant SNPs should remain within the same ballpark as our current analysis (46 out 121,000 SNPs) and the strength of the association (measured with raw p-value) should remain within the same ballpark (padjust min, max and mean).

If our analysis is not powerful enough to allow for 10% genotyping, the number of significant SNPs (and their strength of association) will be outliers.

```{r import misgenotyping data, eval = TRUE, echo = FALSE}
# Evaluating raw p-value distribution from fisher exact test
# https://clauswilke.com/blog/2016/06/13/reading-and-combining-many-tidy-data-files-in-r/

# path to the data
data_path <- "tmp"   

# list the files
files <- dir(data_path, 
             pattern = "2020-10-19-misgenotyping-Fisher-Bonferroni.*.assoc.fisher") 

# load data in a dataframe
data <- data_frame(filename = files) %>% # create a data frame
                                         # holding the file names
  mutate(file_contents = map(filename,          # read files into
           ~ read.csv(file.path(data_path, .), sep = "", header = TRUE)) # a new data column
        ) 

```

```{r compare sig snp numbers, eval = TRUE, echo = FALSE}

#head(data$file_contents[[1]]$P)
# adjust P values for each simulation set

simulation_size <- 1000
sig_snp_num <- c()

for(ticker in 1:simulation_size){
  # adjust for Bonferroni
  adj_pvalue_vec <- p.adjust(data$file_contents[[ticker]]$P,
                             method = "bonferroni")
  
  # save nb of sig SNPs
  sig_snp_num <- c(sig_snp_num, sum(adj_pvalue_vec < 0.05))
  
  
}

# make a dataframe
sig_snps_df <- as.data.frame(sig_snp_num)

##
# plot a histogram of number of significant SNPs
# the real data include 43 sig SNPs. Do we have similar ballpark?

ggplot(sig_snps_df, aes(x = sig_snp_num)) + 
  geom_histogram()
```

```{r compare pval distribution, eval = TRUE, echo = FALSE}
# plot raw p-value distributions
# are they similar than the real distribution?

# https://stackoverflow.com/questions/23758858/how-can-i-extract-elements-from-lists-of-lists-in-r

# make a list of raw p values
raw_p_val_list <- sapply(data$file_contents, `[`, c("P"))
raw_p_val_vec <- unlist(raw_p_val_list)

# make a dataframe
raw_p_val_df <- as.data.frame(raw_p_val_vec)

# plot the distribution of simulated p-values
raw_p_val_df %>%
  ggplot( aes(x=raw_p_val_vec)) +
    geom_density(fill="#69b3a2", color="#e9ecef", alpha=0.8)

# run a Z test or a Kolmogorov-Smirnov test
# for this I need the original p-values (real)
```


