---
title: "Create perfect SNP"
author: "Emeline Favreau"
date: "01/04/2019"
output: pdf_document
---
I copied the first SNP coded in the vcf file. I will replace all genotype codes for homozygote REF for monogynous samples, and homozygote ALT for polygynous samples.

```{r eval = TRUE, echo = FALSE, include = TRUE}
# input data
fakeSNPdf <- read.table(file = "fakeSNP-gt_only", sep = ":")
# add column names
colnames(fakeSNPdf) <- c("GT", "DP", "AD", "RO", "QR", "AO", "QA", "GL")

# import sample names
sample_names_df <- read.table(file = "sample_names.txt")
# add column names
colnames(sample_names_df) <- c("sample_name")

# combine both df (they are ordered already)
GT_info_df <- cbind(fakeSNPdf, sample_names_df)

# change the genotype based on phenotype
GT_info_df$GT <- gsub(pattern = ".*-P", x = sample_names, replacement = "1/1")
GT_info_df$GT <- gsub(pattern = "I27-N", x = sample_names, replacement = "1/1")
GT_info_df$GT <- gsub(pattern = ".*-M", x = sample_names, replacement = "0/0")
GT_info_df$GT <- gsub(pattern = "A56-N", x = sample_names, replacement = "0/0")

# update the missing genotype
GT_info_df$GT[GT_info_df$DP == "."] <- "."

# keep all but the sample names
updated_GT_snp <- subset(GT_info_df, select = -sample_name)

# save output
write.table(x = updated_GT_snp, file = "fakeSNP-gt_only-updated", quote = FALSE, row.names = FALSE, col.names = FALSE, sep = ":")

```
