---
title: "Simulate Solenopsis and Formica SNPs"
author: "Emeline Favreau"
date: "23rd August 2019"
output: pdf_document
---

# Solenopsis model

I copied the first SNP coded in the VCF file. 
I will create a number of SNPs that is realistic, based on Pracana et al 2017 (15,367 fixed difference in supergene out of 628,213 SNPs).
Each monogynous colony will be represented by 0/0 (homozygote for the monogynous reference).
For each polygynous colony, I will draw a worker (BB at 1/3 rate. Bb at 2/3 rate based on Buechel et al 2014), and will assign realistic genotype (BB 0/0, Bb (0/1)).

```{r load data, eval = TRUE, echo = FALSE, include = TRUE}
# input SNP data
SolSNPdf <- read.table(file = "fakeSNPSol-gt_only", sep = ":")

# import sample names
sample_names_df <- read.table(file = "sample_names.txt")

# import SNP header
fakeSNPSolheader <- read.table(file = "snp-first-line", sep = "\t")

# import population info
pop_info_df <- read.csv(file = "S2_pheidole_pop_paper.csv")

```

There are 3,757 real SNPs in this dataset (108 samples, coding, 100% sample support, within-population polymorphic). 
We use Solenopsis data to estimate the number of supergene SNPs that are fixed between social type.

```{r Solenopsis calculate number of SNPs to create, eval = TRUE, echo = FALSE, include = TRUE}

# frequence of supergene fixed difference SNPs
solenopsis_supergene_snp_count <- 15367
solenopsis_total_snp_count <- 628213
solenopsis_non_supergene_snp_count <- solenopsis_total_snp_count - solenopsis_supergene_snp_count

# number of SNPs in the dataset
snp_count <- 3757

# number of SNPs to add
snps_to_add <- round((solenopsis_supergene_snp_count*snp_count)/solenopsis_non_supergene_snp_count)

```

94 SNPs will be simulated.

```{r understand populations diversity, eval = TRUE, echo = FALSE, include = TRUE}
# add column names
colnames(SolSNPdf) <- c("GT", "DP", "AD", "RO", "QR", "AO", "QA", "GL")

# add column names
colnames(sample_names_df) <- c("sample_name")

# change class
pop_info_df$sample.soc <- as.character(pop_info_df$sample.soc)
pop_info_df$population <- as.character(pop_info_df$population)
pop_info_df$gyny <- as.character(pop_info_df$gyny)

# subset pop info for the samples that are in this analysis
pop_info_df_subset <- subset(pop_info_df, subset = sample.soc %in% sample_names_df$sample_name)

# add info about the two samples with missing genotype
pop_info_df_subset <- as.data.frame(rbind(pop_info_df_subset,
                                         c("I27-P", "Italy", "P"),
                                         c("A56-M", "Bruniquel", "M")),
                            stringsAsFactors = FALSE)

# calculate sample counts per population and type
# 69 Bruniquel including 16 monogynous and 53 polygynous
bruniquel_total_sample_count <- nrow(subset(pop_info_df_subset, subset = population == "Bruniquel"))
bruniquel_M_sample_count     <- nrow(subset(pop_info_df_subset, subset = (population == "Bruniquel") & (gyny == "M")))
bruniquel_P_sample_count     <- nrow(subset(pop_info_df_subset, subset = (population == "Bruniquel") & (gyny == "P")))


# 23 Italy including 16 monogynous and 7 polygynous
italy_total_sample_count <- nrow(subset(pop_info_df_subset, subset = population == "Italy"))
italy_M_sample_count     <- nrow(subset(pop_info_df_subset, subset = (population == "Italy") & (gyny == "M")))
italy_P_sample_count     <- nrow(subset(pop_info_df_subset, subset = (population == "Italy") & (gyny == "P")))

# 16 Pyrenees including 5 monogynous and 11 polygynous
pyrenees_total_sample_count <- nrow(subset(pop_info_df_subset, subset = population %in% c("Pyrenees", "Spain")))
pyrenees_M_sample_count     <- nrow(subset(pop_info_df_subset, subset = (population %in% c("Pyrenees", "Spain")) & (gyny == "M")))
pyrenees_P_sample_count     <- nrow(subset(pop_info_df_subset, subset = (population %in% c("Pyrenees", "Spain")) & (gyny == "P")))

```


There are 69 Bruniquel samples, including 16 monogynous and 53 polygynous.
There are 23 Italy samples, including 16 monogynous and 7 polygynous
There are 16 Pyrenees samples, including 5 monogynous and 11 polygynous


```{r prepare data, eval = TRUE, echo = FALSE, include = TRUE}
# combine both df (they are ordered already)
GT_info_df <- cbind(SolSNPdf, sample_names_df)

# change class
GT_info_df$sample_name <- as.character(GT_info_df$sample_name)
GT_info_df$GT <- as.character(GT_info_df$GT)

# create a vector with gyny samples
monogynous_sample_vec <- grep(x = sample_names_df$sample_name, pattern = (".*-M|A56-N"), value = TRUE)
polygynous_sample_vec <- grep(x = sample_names_df$sample_name, pattern = (".*-P|I27-N"), value = TRUE)

# list the type of polygynous worker genotypes
poly_worker_type_vec <- c("0/0", "0/1")

# to vary the samples that have a missing genotype
# loop through the number of SNPs to create
for(snp_position in 1:snps_to_add){
  
    # subset for only the monogynous
    GT_info_df_M <- subset(GT_info_df, subset = sample_name %in% monogynous_sample_vec)
    
    # randomize the sample column
    GT_info_df_M$sample_name <- sample(x = GT_info_df_M$sample_name, size = length(GT_info_df_M$sample_name), replace = FALSE)
    
    # subset for the polygynous
    GT_info_df_P <- subset(GT_info_df, subset = sample_name %in% polygynous_sample_vec)
    
    # randomize the sample column
    GT_info_df_P$sample_name <- sample(x = GT_info_df_P$sample_name, size = length(GT_info_df_P$sample_name), replace = FALSE)
    
    # concatenate monogynous and polygynous
    GT_info_df_sample_randomise <- as.data.frame(rbind(GT_info_df_M, GT_info_df_P),
                                                 stringsAsFactors = FALSE)
    
    # reorder as in the sample_name_df
    GT_info_df_sample_randomise_reordered <- GT_info_df_sample_randomise[match(sample_names_df$sample_name, as.character(GT_info_df_sample_randomise$sample_name)), ]

    # change the genotype based on monogynous phenotype
    GT_info_df_sample_randomise_reordered$GT[GT_info_df_sample_randomise_reordered$sample_name %in% monogynous_sample_vec] <- "0/0"
  
    # sample 71 workers with a probability of 1/3 to be BB (ie 0/0), and 2/3 to be Bb (ie 0/1)
    polygynous_worker_gt <- sample(x = poly_worker_type_vec, size = length(polygynous_sample_vec), replace = TRUE, prob = c(1/3, 2/3))
  
    # change the genotype based on polygynous phenotype
    GT_info_df_sample_randomise_reordered$GT[GT_info_df_sample_randomise_reordered$sample_name %in% polygynous_sample_vec] <- polygynous_worker_gt
  
    # keep all but the sample names
    updated_GT_snp <- subset(GT_info_df_sample_randomise_reordered, select = -sample_name)
    
    # create file name
    file_name <- paste("tmp/updated_GT_snp", snp_position, sep = "")
    
    # save output
    write.table(x = updated_GT_snp, file = file_name,
              quote = FALSE, row.names = FALSE, col.names = FALSE, 
              sep = ":", eol = "\t")
  
    # update the SNP header
    fakeSNPSolheader$V1 <- gsub(x = fakeSNPSolheader$V1,
                                  pattern = "contig_01|fake_[0-9]+",
                                  replacement = paste("fake_", snp_position, sep = ""))
    
    # create file name
    file_name <- paste("tmp/updated_GT_snp_header", snp_position, sep = "")
    
    
    # save output
    write.table(x = fakeSNPSolheader, file = file_name,
              quote = FALSE, row.names = FALSE, col.names = FALSE, 
              sep = "\t")
}





```

Here we generate 94 fake SNPs for which all monogynous samples and 1/3 polygynous samples are homozygote for the reference, and 2/3 polygynous samples are heterozygote.


# Formica selysi
There are 3,757 real SNPs in this dataset (108 samples, coding, 100% sample support, within-population polymorphic). 
We use _Formica selysi_ knowledge to estimate the number of supergene SNPs that are fixed between social type (Purcell et al 2014).

In this system, the monogynous workers are always homozygotes for the reference.
The polygynous workers are either heterozygotes (freq: 0.68) or homozygote for alternative (freq: 0.32).


```{r Formica calculate number of SNPs to create, eval = TRUE, echo = FALSE, include = TRUE}

# frequence of supergene fixed difference SNPs
formica_supergene_snp_count <- 643
formica_total_snp_count <- 18199
formica_non_supergene_snp_count <- formica_total_snp_count - formica_supergene_snp_count

# number of SNPs in the dataset
snp_count <- 3757

# number of SNPs to add
snps_to_add <- round((formica_supergene_snp_count*snp_count)/formica_non_supergene_snp_count)

```

138 SNPs will be simulated.

```{r Formica prepare data, eval = TRUE, echo = FALSE, include = TRUE}
# combine both df (they are ordered already)
GT_info_df <- cbind(SolSNPdf, sample_names_df)

# change class
GT_info_df$sample_name <- as.character(GT_info_df$sample_name)
GT_info_df$GT <- as.character(GT_info_df$GT)

# create a vector with gyny samples
monogynous_sample_vec <- grep(x = sample_names_df$sample_name, pattern = (".*-M|A56-N"), value = TRUE)
polygynous_sample_vec <- grep(x = sample_names_df$sample_name, pattern = (".*-P|I27-N"), value = TRUE)

# list the type of polygynous worker genotypes
poly_worker_type_vec <- c("0/1", "1/1")

# to vary the samples that have a missing genotype
# loop through the number of SNPs to create
for(snp_position in 1:snps_to_add){
    
    # subset for only the monogynous
    GT_info_df_M <- subset(GT_info_df, subset = sample_name %in% monogynous_sample_vec)
    
    # randomize the sample column
    GT_info_df_M$sample_name <- sample(x = GT_info_df_M$sample_name, size = length(GT_info_df_M$sample_name), replace = FALSE)
    
    # subset for the polygynous
    GT_info_df_P <- subset(GT_info_df, subset = sample_name %in% polygynous_sample_vec)
    
    # randomize the sample column
    GT_info_df_P$sample_name <- sample(x = GT_info_df_P$sample_name, size = length(GT_info_df_P$sample_name), replace = FALSE)
    
    # concatenate monogynous and polygynous
    GT_info_df_sample_randomise <- as.data.frame(rbind(GT_info_df_M, GT_info_df_P),
                                                 stringsAsFactors = FALSE)
    
    # reorder as in the sample_name_df
    GT_info_df_sample_randomise_reordered <- GT_info_df_sample_randomise[match(sample_names_df$sample_name, as.character(GT_info_df_sample_randomise$sample_name)), ]

    # change the genotype based on monogynous phenotype
    GT_info_df_sample_randomise_reordered$GT[GT_info_df_sample_randomise_reordered$sample_name %in% monogynous_sample_vec] <- "0/0"

    # sample 71 workers with a probability of 1/3 to be BB (ie 0/0), and 2/3 to be Bb (ie 0/1)
    polygynous_worker_gt <- sample(x = poly_worker_type_vec, size = length(polygynous_sample_vec), replace = TRUE, prob = c(0.68, 0.32))

    # change the genotype based on polygynous phenotype
    GT_info_df_sample_randomise_reordered$GT[GT_info_df_sample_randomise_reordered$sample_name %in% polygynous_sample_vec] <- polygynous_worker_gt

    # keep all but the sample names
    updated_GT_snp <- subset(GT_info_df_sample_randomise_reordered, select = -sample_name)
    
    # update the SNP header
    fakeSNPSolheader$V1 <- gsub(x = fakeSNPSolheader$V1,
                                pattern = "fake_[0-9]+_",
                                replacement = paste("fake_", snp_position, "_", sep = ""))
    # create file name
    file_name <- paste("tmp/formica_updated_GT_snp_header", snp_position, sep = "")
    
    # save output in tmp
    write.table(x = fakeSNPSolheader, file = file_name,
            quote = FALSE, row.names = FALSE, col.names = FALSE, 
            sep = "\t")
    
    # create file name
    file_name <- paste("tmp/formica_updated_GT_snp", snp_position, sep = "")
    
    # save output in tmp
    write.table(x = updated_GT_snp, file = file_name,
            quote = FALSE, row.names = FALSE, col.names = FALSE, 
            sep = ":", eol = "\t")
}





```

Here we generate 138 fake SNPs based on Formica system, for which all monogynous samples are homozygotes reference, 0.68 polygynous are heterozygote, 0.32 are homozygote alternative.