---
title: "2019-03-11-checking-busco-results"
author: "Emeline Favreau"
date: "11/03/2019"
output: pdf_document
---

I blasted the query Solenopsis single-copy ortholog genes against _Pheidole_ MinION-flye assembly to investigate if there are some genes spanning the extremes of two MinION-flye contigs? If so, put the two sequences together.


###############################################################################

```{r eval = TRUE, echo = FALSE, include = TRUE}
# input the blast input
raw_blast_output <- read.table(file = "result/MinION-flye-solenopsis-genes-output-blast.txt",
                               header = FALSE)
colnames(raw_blast_output) <- c("qseqid", "sseqid", "pident", "length",
                                "qstart", "qend", "sstart", "send",
                                "evalue", "bitscore", "qlen", "qcovs")
raw_blast_output$sseqid <- as.character(raw_blast_output$sseqid)
raw_blast_output$qseqid <- as.character(raw_blast_output$qseqid)

```


If a gene spans two contigs, the query will not be 100% covered by one contig only.
Maybe I could see this signal by checking the summary of qcovs by qseqid?
Expectation: low coverage of a gene per contig possible equals to gene split between contigs?

```{r eval = TRUE, echo = FALSE, include = TRUE}
# list of all solenopsis genes
sol_genes_vec <- unique(as.character(raw_blast_output$qseqid))

# "There are 4240 Solenopsis single-copy ortholog genes in this analysis."
print(paste("There are", length(sol_genes_vec), "Solenopsis single-copy ortholog genes in this analysis."))

# prepare a table to obtain name of Sol gene and mean of cover query
query_cover_summary_df <- matrix(NA, nrow = length(sol_genes_vec), ncol = 2)

# fill the table
for(position in 1:length(sol_genes_vec)){
  # subset blast outpout for just one Solenopsis gene
  subset_df <- raw_blast_output[raw_blast_output$qseqid == sol_genes_vec[position], ]
  # calculate the mean coverage of the gene by the MinION sequences
  query_cover_mean <- mean(subset_df$qcovs)
  # add result in table
  query_cover_summary_df[position, ] <- c(sol_genes_vec[position], query_cover_mean)
}

# change to df class
query_cover_summary_df <- as.data.frame(query_cover_summary_df)
# name the columns
colnames(query_cover_summary_df) <- c("gene_name", "query_cover")
# change to character
query_cover_summary_df$gene_name <- as.character(query_cover_summary_df$gene_name)

# take the extreme lowest average value of query cover
query_cover_summary_df$query_cover <- as.numeric(as.character(query_cover_summary_df$query_cover))
# order the df to show the lowest values on top
query_cover_summary_df <- query_cover_summary_df[order(query_cover_summary_df$query_cover), ]


test <- raw_blast_output[raw_blast_output$qseqid == query_cover_summary_df$gene_name[2], ]
```



# Conclusion:
The MinION-flye assembly probably has more genes present than BUSCO predicted (42% complete orthologs).


