---
title: "fasta_headers_to_taxa"
author: "Catherine Okuboyejo"
date: "11 July 2019"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Load libraries; install from scratch if needed
basic_libraries <- c("stringr", "dplyr")
for (lib in basic_libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                install.packages(lib)
                library(lib, character.only = TRUE )
        }
}

# if (!require("stringr", character.only = TRUE)) {
#         install.packages(c("stringr", "Dplyr"), repos = 'https://cran.ma.imperial.ac.uk/')
#         library(stringr, character.only = TRUE)
#         library(Dplyr, character.only = TRUE)
#          
# }
# additional setup
source("phylum_from_NCBI_taxid.R")
Sys.setenv(ENTREZ_KEY = 'b1b813d18133d806850967304bb99f977007') # SET NCBI ENTREZ KEY HERE IF YOU HAVE ONE (code still works without)


# load size of contigs
contig_length <- read.table("Ppal_E.contig.length") 
```


```{r get taxonomy data for all hits, echo = FALSE, warning = FALSE}
# read in headers from protein sequence of uniref hits
hits_unique <- read.table("tmp/Ppal_E.diamond.headers", sep = "%", 
                          quote = "", stringsAsFactors = FALSE)

# put relevant information into columns
hits_unique$uniref_id <- str_extract(hits_unique$V1, "(?<=^>)\\S+")
hits_unique$V1 <- str_replace(hits_unique$V1, "^>\\S+\\s", "")
hits_unique$cluster_name <- str_extract(hits_unique$V1, "^.+(?=\\sn=)")
hits_unique$taxid <- str_extract(hits_unique$V1, "(?<=TaxID=)\\d+")
hits_unique$rep_id <- str_extract(hits_unique$V1, "(?<=RepID=)\\S+")
hits_unique <- select(hits_unique, -V1)

# look up taxids to get phylum and class
# to avoid overloading ncbi, take only unique taxids to look up
tax_id_to_tax_table <- data.frame(taxid = unique(hits_unique$taxid), 
                                  phylum = NA, class = NA, order = NA, other = NA)

# get taxa
for (i in 1:nrow(tax_id_to_tax_table)) {
        current_id <- tax_id_to_tax_table$taxid[i]
        tax_id_to_tax_table[i, c("phylum", "class", "order", "other")] <- suppressWarnings(ranks_from_NCBI_taxid(current_id))
        if (length(Sys.getenv("ENTREZ_KEY")) == 0) Sys.sleep(time = 1) # slow down to avoid too many requests error if there is no entrez key
}
```

```{r add taxonomy data to each hit, echo = FALSE}
# read in first two columns of diamond output to match uniref hits to contigs
diamond_output <- read.table("tmp/Ppal_E.diamond", 
                             col.names = c("ref_name", "uniref_id", "pident",
                                           "length", "mismatch", "gapopen", 
                                           "qstart", "qend", "sstart", "send",
                                           "evalue", "bitscore"),
                             stringsAsFactors = FALSE)

# merge diamond output with unique hits to match contig with uniref hits
suppressWarnings(scaf_uniref_hits <- right_join(diamond_output, hits_unique, by = "uniref_id"))

# add phyla from taxid table
suppressWarnings(scaf_uniref_hits <- left_join(scaf_uniref_hits, tax_id_to_tax_table, by = "taxid"))

# save table
write.table(scaf_uniref_hits, file = "tmp/Ppal_E.diamond.scaf_uniref_hits", 
            quote = FALSE, row.names = FALSE)
```

```{r understand output, echo = FALSE}
# order dataframe by name of contig
scaf_uniref_hits_ordered <- scaf_uniref_hits[order(scaf_uniref_hits$ref_name), ]

# add columns
colnames(contig_length) <- c("contig", "contig_length")

# add length of contig
scaf_uniref_hits_ordered$contig_length <- contig_length$contig_length[match(scaf_uniref_hits_ordered$ref_name, contig_length$contig)]
  
# are there any contig with one hit only?
temp_file <- sort(table(scaf_uniref_hits_ordered$ref_name))
contigs_with_unique_hits <- names(temp_file[temp_file == 1])
# n = 83

# keep contigs with unique hit if they are Hymenoptera (n = 72)
hymenoptera_Ppal_E_contigs <- unlist(subset(scaf_uniref_hits_ordered, subset = ref_name %in% contigs_with_unique_hits & order == "Hymenoptera", select = ref_name))

# those contigs with unique hit and not Hymenoptera, they are less than 205bp so not interesting
# subset(scaf_uniref_hits_ordered, subset = ref_name %in% contigs_with_unique_hits & order != "Hymenoptera", select = c(ref_name, length, order))
not_hymenoptera_Ppal_E_contigs_small <- unlist(subset(scaf_uniref_hits_ordered, subset = ref_name %in% contigs_with_unique_hits & order != "Hymenoptera", select = ref_name))

# 3 contigs with unique hit and uncharacterised order are very small (less than 220bp) so not so interesting
not_hymenoptera_Ppal_E_contigs_small <- c(not_hymenoptera_Ppal_E_contigs_small, 
                                          unlist(subset(scaf_uniref_hits_ordered,
                                                        subset = ref_name %in% contigs_with_unique_hits & is.na(order), select = ref_name)))



# keep names of contigs in a vector
contig_vec <- unique(scaf_uniref_hits_ordered$ref_name)
# make a list for each contig
taxa_hit_list <- list()
# loop through each contig
for(position in 1:length(contig_vec)){
  # subset to this contig
  scaf_uniref_hits_ordered_subset <- subset(scaf_uniref_hits_ordered, subset = ref_name == contig_vec[position])
  # keep taxon orders
  taxa_hit_list[[position]] <- unique(scaf_uniref_hits_ordered_subset$order)
}

# name elements
names(taxa_hit_list) <- contig_vec

# are there some contigs that have multiple hits, but none of them are Hymenoptera?
# keep contigs that are Hymenoptera
contigs_one_hit_hymenoptera <- c()

# keep contigs that have one hit that is not Hymenoptera
contigs_one_hit_not_hymenoptera <- c()

# keep contigs that have more than one hit 
contigs_many_hits <- c()

# loop through each contig
for(position in 1:length(contig_vec)){
  # count number of hits
  number_of_hits <- length(taxa_hit_list[[position]])
  # keep the contig if there is one hit and it is hymenoptera
  if(number_of_hits == 1){
    if(taxa_hit_list[[position]] %in% "Hymenoptera"){
      # keep name of contigs
      contigs_one_hit_hymenoptera <- c(contigs_one_hit_hymenoptera, contig_vec[position])
    } else {
      #print("one taxon, but not Hymenoptera")
      # keep name of contig
      contigs_one_hit_not_hymenoptera <- c(contigs_one_hit_not_hymenoptera, contig_vec[position])
  }
    } else {
      #print("more than one taxon")
      # keep name of contig
      contigs_many_hits <- c(contigs_many_hits, contig_vec[position])
  }
}


# create a vector with the names of contigs that have at least one hymenoptera hit
contigs_many_hits_contain_hymenoptera <- c()
# create a vector with the names of contigs that do not have hymenoptera hit
contigs_many_hits_no_hymenoptera <- c()
# loop through the contigs that have multiple order hits
for(position in 1:length(contigs_many_hits)){
  # if the contig contains at least once the order hit "Hymenoptera"
  if(nrow(subset(scaf_uniref_hits_ordered, subset = ref_name %in% contigs_many_hits[position] & order == "Hymenoptera")) > 0){
    # add the name of contig to the vector
    contigs_many_hits_contain_hymenoptera <- c(contigs_many_hits_contain_hymenoptera, contigs_many_hits[position])
  } else {
    contigs_many_hits_no_hymenoptera <- c(contigs_many_hits_no_hymenoptera, contigs_many_hits[position])
  }
}

# investigate the multiple hits that are not hymenoptera
test <- subset(scaf_uniref_hits_ordered, subset = ref_name %in% contigs_many_hits_no_hymenoptera)

# investigate length
length_proportion_df <- as.data.frame(cbind(
                                          c(hymenoptera_Ppal_E_contigs,
                                          not_hymenoptera_Ppal_E_contigs_small,
                                          contigs_many_hits_contain_hymenoptera,
                                          contigs_many_hits_no_hymenoptera),
                                         
                                          c(contig_length$contig_length[match(hymenoptera_Ppal_E_contigs,            contig_length$contig)],
                                            contig_length$contig_length[match(not_hymenoptera_Ppal_E_contigs_small,  contig_length$contig)],
                                            contig_length$contig_length[match(contigs_many_hits_contain_hymenoptera, contig_length$contig)],
                                            contig_length$contig_length[match(contigs_many_hits_no_hymenoptera,      contig_length$contig)]),
                                        
                                            c(rep("one_hit_hym",      times = length(hymenoptera_Ppal_E_contigs)),
                                              rep("one_hit_no_hym",   times = length(not_hymenoptera_Ppal_E_contigs_small)),
                                              rep("multi_hits_hym",   times = length(contigs_many_hits_contain_hymenoptera)),
                                              rep("multi_hit_no_hym", times = length(contigs_many_hits_no_hymenoptera)))),
                 stringsAsFactors = FALSE, row.names = NULL
)

# name columns
colnames(length_proportion_df) <- c("contig", "length", "diamond_category")

# calculate the proportion of the genome per diamond category
one_hit_hym_length <- subset(length_proportion_df, subset = diamond_category == "one_hit_hym", select = length)
# 0.0871837
#sum(as.numeric(one_hit_hym_length$length)) / 270000000 * 100

# calculate the proportion of the genome per diamond category
one_hit_no_hym_length <- subset(length_proportion_df, subset = diamond_category == "one_hit_no_hym", select = length)
# 0.01
#sum(as.numeric(one_hit_no_hym_length$length)) / 270000000 * 100

# calculate the proportion of the genome per diamond category
multi_hits_hym_length <- subset(length_proportion_df, subset = diamond_category == "multi_hits_hym", select = length)
# 94.26948
#sum(as.numeric(multi_hits_hym_length$length)) / 270000000 * 100

# calculate the proportion of the genome per diamond category
multi_hit_no_hym_length <- subset(length_proportion_df, subset = diamond_category == "multi_hit_no_hym", select = length)
# 1.177403
#sum(as.numeric(multi_hit_no_hym_length$length)) / 270000000 * 100

# save contigs that are Hymenoptera
all_hymenoptera_Ppal_E_contigs <- c(hymenoptera_Ppal_E_contigs, contigs_many_hits_contain_hymenoptera)
write.table(all_hymenoptera_Ppal_E_contigs, file = "result/all_hymenoptera_Ppal_E_contigs", row.names = FALSE, quote = FALSE, col.names = FALSE)
```

There are 3,251 contigs in Ppal_E, 2,608 are Hymenoptera and represent 94% of Ppal_E length.

Name of object  |  Number of contigs  | Content | Proportion of genome (%)
------------- |  ------------- | ------------- | -------------
contigs_many_hits_contain_hymenoptera | 2,536 | multiple hits, at least 1 Hymenoptera | 94.27
hymenoptera_Ppal_E_contigs |  72 | single hit, Hymenoptera | 0.087
contigs_many_hits_no_hymenoptera | 62 | multiple hits, no Hymenoptera | 1.18
not_hymenoptera_Ppal_E_contigs_small | 11 | single hit, not Hymenoptera | 0.01

