---
title: "Pheidole read coverage by social type"
author: "Emeline Favreau"
date: "05/06/2019"
output: pdf_document
---

I investigate the differences between social type (monogynous, polygynous) at various levels: 

- raw read coverage, 
- CG content, 
- proportion of read mapped to the monogynous reference, 
- mapping quality.

```{r setup, include = FALSE}
#knitr::opts_chunk$set(echo = TRUE, fig.width = 10, fig.height = 8)
```


```{r import files and load libraries, include = FALSE}
# import raw sequence coverage
raw_sequence_coverage <- read.csv("../../2018-05-30-pheidole-raw-115samples/results/2018-06-03-sample-coverage/115_pheidole_coverage.txt", header = FALSE, sep = "\t", stringsAsFactors = FALSE)

# import qualimap stats
qualimap_stats <- read.csv("2019-03-15-map_reads_to_minion_flye/all-samples-qualimap-stats", header = FALSE, sep = "\t", stringsAsFactors = FALSE)

# import metadata
metadata_df <- read.csv("2019-04-03-coding-only-bruniquel-maf10percent/S2_pheidole_pop_paper.csv", header = TRUE, sep = ",", stringsAsFactors = FALSE)

# get and/or load all the libraries
basic_libraries <- c("ggplot2", "tidyr")
for (lib in basic_libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                install.packages(lib)
                library(lib, character.only = TRUE )
        }
}
```

## Sample coverage by the raw reads

Here I investigate if there is a difference of coverage by raw sequences between samples of monogynous origin and polygynous origin.

```{r raw sequences coverage, eval = TRUE, echo = FALSE, warning = FALSE}
## Step 1: tidy input dataframe
# name columns
colnames(raw_sequence_coverage) <- c("sample_name", "sequence_count")

# update name of samples (due to raw samples mis-named)
# I27-P, A56-M, andrea-P, muna-P
raw_sequence_coverage$sample_name <- gsub(pattern = "I27\\-N.*",    x = raw_sequence_coverage$sample_name, replacement = "I27-P_R1_001")
raw_sequence_coverage$sample_name <- gsub(pattern = "A56\\-N.*",    x = raw_sequence_coverage$sample_name, replacement = "A56-M_R1_001")
raw_sequence_coverage$sample_name <- gsub(pattern = "andrea\\-N.*", x = raw_sequence_coverage$sample_name, replacement = "andrea-P_R1_001")
raw_sequence_coverage$sample_name <- gsub(pattern = "muna\\-N.*",   x = raw_sequence_coverage$sample_name, replacement = "muna-P_R1_001")

# tidy the dataframe: the sequence count in second line equals the sequence count in first line
for(position in seq(1, nrow(raw_sequence_coverage)-1, by = 2)){
  # update the sequence count
  raw_sequence_coverage[position, 2] <- raw_sequence_coverage[position+1, 2]
}
# remove the lines with "Total Sequences""
raw_sequence_coverage_cleaned <- subset(raw_sequence_coverage, subset = sample_name != "Total Sequences")





## Step 2: extract information for each sample 

# make vector of colony names
colony_name_vec <- unique(gsub(pattern = "\\_R.*", x = raw_sequence_coverage_cleaned$sample_name, replacement = ""))
# make dataframe to gather information for each sample
sequence_coverage_per_social <- as.data.frame(matrix(ncol = 3, nrow = length(colony_name_vec)))
# name the columns
colnames(sequence_coverage_per_social) <- c("sample_name", "total_sequence_count", "social")
# add the colony names
sequence_coverage_per_social$sample_name <- colony_name_vec
# add the social type (M or P)
sequence_coverage_per_social$social <- gsub(pattern = ".*\\-", x = sequence_coverage_per_social$sample_name, replacement = "")

# looping through one sample at a time
for(position in 1:length(colony_name_vec)){
  # subset the raw df for this sample
  # add the sum of sequence count for R1 and R2
  sequence_coverage_per_social$total_sequence_count[sequence_coverage_per_social$sample_name == colony_name_vec[position]] <- sum(raw_sequence_coverage_cleaned$sequence_count[grep(pattern = colony_name_vec[position], x = raw_sequence_coverage_cleaned$sample_name)])
}





## Step 3: Are those distributions statistically different?
# test of the equality of continuous probability distribution used to compare the two social groups
# run a Kolmogorov-Smirnov test
# a.k.a testing the shape of the distribution
# The null hypothesis is that both groups were sampled from populations with identical distributions
total_sequence_countM <- sequence_coverage_per_social$total_sequence_count[sequence_coverage_per_social$social== "M"]
total_sequence_countP <- sequence_coverage_per_social$total_sequence_count[sequence_coverage_per_social$social== "P"]
#ks.test(x = total_sequence_countM, y = total_sequence_countP)$p.value
# 0.005581148 < 0.05, significant, the two distributions shapes are different
#print(paste("KS test p =", ks.test(x = total_sequence_countM, y = total_sequence_countP)$p.value))
# test of the null hypothesis that it is equally likely that a randomly selected value from one sample will be less than or greater than a randomly selected value from a second sample
# run a Mann–Whitney U test (wilcoxon) test
# testing if the medians are different (non-paramteric ttest)
#wilcox.test(x = total_sequence_countM, y = total_sequence_countP)$p.value
# 0.01008714 < 0.05, significant, the two distributions median are from different distributions

# # Carlos hinted that the outlier E115 (purple peak on the right) might affect the tests
# # removing the outlier "E15-P" and running the test again
# sequence_coverage_per_social_NoOutlier <- subset(sequence_coverage_per_social, subset = sample_name != "E15-P")
# total_sequence_countM <- sequence_coverage_per_social_NoOutlier$total_sequence_count[sequence_coverage_per_social_NoOutlier$social== "M"]
# total_sequence_countP <- sequence_coverage_per_social_NoOutlier$total_sequence_count[sequence_coverage_per_social_NoOutlier$social== "P"]
# ks.test(x = total_sequence_countM, y = total_sequence_countP)$p.value
# # 0.00413318 < 0.05, significant, the two distributions shapes are different
# wilcox.test(x = total_sequence_countM, y = total_sequence_countP)$p.value
# # 0.006575152 < 0.05, significant, the two distributions median are from different distributions



## Step 4: plot the results
plot_title    <- "Figure 1: Read coverage distribution by social type"
plot_subtitle <- "Data from FastQC count of Total Sequences"
x_title       <- "Sequence count"
annotation_text <- paste("KS test p =", round(ks.test(x = total_sequence_countM, y = total_sequence_countP)$p.value, digits = 5),
                         "\nWilcoxon test p =", round(wilcox.test(x = total_sequence_countM, y = total_sequence_countP)$p.value, digits = 5))
ggplot(sequence_coverage_per_social, aes(x = total_sequence_count, fill = social, colour = social)) +
   geom_density(alpha = 0.5) +
   labs(title = plot_title, subtitle = plot_subtitle, x = x_title) +
   # Monogynous orange f, Polygynous purple 9
   scale_fill_manual(values   = c("#f1a340", "#998ec3")) +
   scale_colour_manual(values = c("#f1a340", "#998ec3")) +
   theme_classic() +
   annotate("text", x = 50000000, y = 0.00000007, label = annotation_text)




```

The two distributions are not similar in shape nor in median.
The purple peak on the right is outlier E14 (ca. 66x coverage). 
Even if I remove this outlier, the distributions are statistically different.

The difference between the two sample groups can be explained by: 

+ randomness, 
+ batch effect in the lab (very unlikely), 
+ P workers being smaller (less DNA -> library prep concentration slightly smaller? Unlikely: before DNA conc normalisation, the P samples had slightly higher conc ), 
+ enzyme fragmenting differently between social group, 
+ PCR biased toward M workers?

It will be difficult to disentangle the impacts of everything. We will keep in mind this difference.

## GC content

```{r gc percentage, eval = TRUE, echo = FALSE, warning = FALSE}
## Step 1: tidy input dataframe
# name columns
colnames(qualimap_stats) <- c("sample_name", "mapped_reads_percent", "mapped_reads_count", "mean_mapping_quality", "gc_percentage", "mismatches_number")
# update name of samples (due to raw samples mis-named)
# I27-P, A56-M, andrea-P, muna-P
qualimap_stats$sample_name <- gsub(pattern = "I27\\-N.*",    x = qualimap_stats$sample_name, replacement = "I27-P")
qualimap_stats$sample_name <- gsub(pattern = "A56\\-N.*",    x = qualimap_stats$sample_name, replacement = "A56-M")
qualimap_stats$sample_name <- gsub(pattern = "andrea\\-N.*", x = qualimap_stats$sample_name, replacement = "andrea-P")
qualimap_stats$sample_name <- gsub(pattern = "muna\\-N.*",   x = qualimap_stats$sample_name, replacement = "muna-P")
# add the social type (M or P)
qualimap_stats$social <- gsub(pattern = ".*\\-", x = qualimap_stats$sample_name, replacement = "")
# change to factor to order the x axis in ggplot 
qualimap_stats$sample_name <- factor(qualimap_stats$sample_name, levels = qualimap_stats$sample_name[order(qualimap_stats$social)])

# test for differences in means
if(t.test(
  x = qualimap_stats$gc_percentage[qualimap_stats$social == "M"],
  y = qualimap_stats$gc_percentage[qualimap_stats$social == "P"])$p.value < 0.05){
  print(paste("GC content is significantly different between monogynous and polygynous samples,\n t.test p-value = ", round(t.test(x = qualimap_stats$gc_percentage[qualimap_stats$social == "M"],
       y = qualimap_stats$gc_percentage[qualimap_stats$social == "P"])$p.value, digits = 3)))
} else {
  print(paste("GC content is not significantly different between monogynous and polygynous samples, \n t.test p-value = ", round(t.test(x = qualimap_stats$gc_percentage[qualimap_stats$social == "M"],
       y = qualimap_stats$gc_percentage[qualimap_stats$social == "P"])$p.value, digits = 3)))
}

## Step 2: plot gc percentage per social type
plot_title    <- paste("Figure 2: GC content proportion by social type")
plot_subtitle <- paste("Data from Qualimap report")

ggplot(data = qualimap_stats, aes(x = sample_name, y = gc_percentage, fill = social)) +
    geom_bar(colour = "black", stat = "identity") +
    guides(fill = FALSE) +
    labs(title = plot_title, subtitle = plot_subtitle) +
    # Monogynous orange f, Polygynous purple 9
    scale_fill_manual(values   = c("#f1a340", "#998ec3")) +
    scale_colour_manual(values = c("#f1a340", "#998ec3")) +
    theme_classic() + 
    theme(axis.text.x  = element_text(angle = 45, vjust = 0.5, size = 4.5))
```

GC content is different between social type.




## Proportion of mapped reads

```{r mapped reads, eval = TRUE, echo = FALSE, warning = FALSE}
## plot mapped reads per social type
plot_title    <- paste("Figure 3: Mapped read proportion by social type")
plot_subtitle <- paste("Data from Qualimap report")

# test for differences in means
print(paste("Proportion of mapped reads is not significantly different between monogynous and polygynous samples, t.test p-value = ", t.test(x = qualimap_stats$mapped_reads_percent[qualimap_stats$social == "M"],
       y = qualimap_stats$mapped_reads_percent[qualimap_stats$social == "P"])$p.value)) # 0.583597274243485

ggplot(data = qualimap_stats, aes(x = sample_name, y = mapped_reads_percent, fill = social)) +
    geom_bar(colour = "black", stat = "identity") +
    guides(fill = FALSE) +
    labs(title = plot_title, subtitle = plot_subtitle) +
    # Monogynous orange f, Polygynous purple 9
    scale_fill_manual(values   = c("#f1a340", "#998ec3")) +
    scale_colour_manual(values = c("#f1a340", "#998ec3")) +
    theme_classic() + 
    theme(axis.text.x  = element_text(angle = 45, vjust = 0.5, size = 4.5))




## plot mapped reads count per social type
plot_title    <- paste("Figure 3b: Mapped read count by social type")
plot_subtitle <- paste("Data from Qualimap report")

# test for differences in means
print(paste("Mapped reads count is not significantly different between monogynous and polygynous samples, t.test p-value = ", t.test(x = qualimap_stats$mapped_reads_count[qualimap_stats$social == "M"],
       y = qualimap_stats$mapped_reads_count[qualimap_stats$social == "P"])$p.value)) # 0.4046523

ggplot(data = qualimap_stats, aes(x = sample_name, y = mapped_reads_count, fill = social)) +
    geom_bar(colour = "black", stat = "identity") +
    guides(fill = FALSE) +
    labs(title = plot_title, subtitle = plot_subtitle) +
    # Monogynous orange f, Polygynous purple 9
    scale_fill_manual(values   = c("#f1a340", "#998ec3")) +
    scale_colour_manual(values = c("#f1a340", "#998ec3")) +
    theme_classic() + 
    theme(axis.text.x  = element_text(angle = 45, vjust = 0.5, size = 4.5))



## plot unmapped reads count per social type
plot_title    <- paste("Figure 3c: Non-mapped read count by social type")
plot_subtitle <- paste("Data from Qualimap report")

# update class
qualimap_stats$sample_name <- as.character(qualimap_stats$sample_name)

# add total amount of reads 
qualimap_stats$total_sequence_count <- sequence_coverage_per_social$total_sequence_count[match(qualimap_stats$sample_name, sequence_coverage_per_social$sample_name)]

# calculate non-mapped read count
qualimap_stats$non_mapped_read_count <- qualimap_stats$total_sequence_count - qualimap_stats$mapped_reads_count


# change to factor to order the x axis in ggplot 
qualimap_stats$sample_name <- factor(qualimap_stats$sample_name, levels = qualimap_stats$sample_name[order(qualimap_stats$social)])

# test for differences in means
print(paste("Non-mapped reads count is not significantly different between monogynous and polygynous samples, t.test p-value = ", t.test(x = qualimap_stats$non_mapped_read_count[qualimap_stats$social == "M"],
       y = qualimap_stats$non_mapped_read_count[qualimap_stats$social == "P"])$p.value)) # 0.65439

ggplot(data = qualimap_stats, aes(x = sample_name, y = non_mapped_read_count, fill = social)) +
    geom_bar(colour = "black", stat = "identity") +
    guides(fill = FALSE) +
    labs(title = plot_title, subtitle = plot_subtitle) +
    # Monogynous orange f, Polygynous purple 9
    scale_fill_manual(values   = c("#f1a340", "#998ec3")) +
    scale_colour_manual(values = c("#f1a340", "#998ec3")) +
    theme_classic() + 
    theme(axis.text.x  = element_text(angle = 45, vjust = 0.5, size = 4.5))






## plot unmapped + mapped reads count per social type
plot_title    <- paste("Figure 3d: Read count by social type and by mapping status")
plot_subtitle <- paste("Data from Qualimap report")

# subset to info needed
qualimap_stats_wide <- subset(qualimap_stats,
                            select = c(sample_name, mapped_reads_count, social, non_mapped_read_count))

# transform dataset long format
qualimap_stats_lf <- pivot_longer(data = qualimap_stats_wide,
                                  cols = c(mapped_reads_count, non_mapped_read_count), 
                                  names_to = "read_type",
                                  names_repair = "check_unique",
                                  values_to = "read_count")
 
# Stacked + percent
ggplot(data = qualimap_stats_lf, aes(fill = read_type, y = read_count, x = sample_name)) + 
    geom_bar(colour = "black", stat = "identity", position = "fill") +
    guides(fill = FALSE) +
    labs(title = plot_title, subtitle = plot_subtitle) +
    # Monogynous orange f, Polygynous purple 9
    #scale_fill_manual(values   = c("#f1a340", "#998ec3")) +
    #scale_colour_manual(values = c("#f1a340", "#998ec3")) +
    theme_classic() + 
    theme(axis.text.x  = element_text(angle = 45, vjust = 0.5, size = 4.5))

```

The proportion of reads that mapped to the reference does not differ much between social type.


## Mapping quality

```{r mean mapping quality, eval = TRUE, echo = FALSE, warning = FALSE}
## Step 1: Are those distributions statistically different?
# test of the equality of continuous probability distribution used to compare the two social groups
# run a Kolmogorov-Smirnov test
# a.k.a testing the shape of the distribution
# The null hypothesis is that both groups were sampled from populations with identical distributions
mean_mapping_qualityM <- qualimap_stats$mean_mapping_quality[qualimap_stats$social== "M"]
mean_mapping_qualityP <- qualimap_stats$mean_mapping_quality[qualimap_stats$social== "P"]
#ks.test(x = mean_mapping_qualityM, y = mean_mapping_qualityP)$p.value
# 0.002136609 < 0.05, significant, the two distributions shapes are different
# print(paste("KS test p =", ks.test(x = total_sequence_countM, y = total_sequence_countP)$p.value))
# test of the null hypothesis that it is equally likely that a randomly selected value from one sample will be less than or greater than a randomly selected value from a second sample
# run a Mann–Whitney U test (wilcoxon) test
# testing if the medians are different (non-paramteric ttest)
#wilcox.test(x = mean_mapping_qualityM, y = mean_mapping_qualityP)$p.value
# 0.0003633979 < 0.05, significant, the two distributions median are from different distributions

# # Carlos hinted that the outlier E115 (purple peak on the right) might affect the tests
# # removing the outlier "E15-P" and running the test again
#qualimap_stats_NoOutlier <- subset(qualimap_stats, subset = sample_name != "E15-P")
#mean_mapping_qualityM <- qualimap_stats_NoOutlier$mean_mapping_quality[qualimap_stats_NoOutlier$social== "M"]
#mean_mapping_qualityP <- qualimap_stats_NoOutlier$mean_mapping_quality[qualimap_stats_NoOutlier$social== "P"]
# ks.test(x = total_sequence_countM, y = total_sequence_countP)$p.value
# # 0.001548038 < 0.05, significant, the two distributions shapes are different
# wilcox.test(x = total_sequence_countM, y = total_sequence_countP)$p.value
# # 0.0002285013 < 0.05, significant, the two distributions median are from different distributions



## plot mean mapping quality per social type
plot_title    <- paste("Figure 4: Mean mapping quality by social type")
plot_subtitle <- paste("Data from Qualimap report")
annotation_text <- paste("KS test p =", round(ks.test(x = mean_mapping_qualityM, y = mean_mapping_qualityP)$p.value, digits = 5),
                         "\nWilcoxon test p =", round(wilcox.test(x = mean_mapping_qualityM, y = mean_mapping_qualityP)$p.value, digits = 5))
ggplot(qualimap_stats, aes(x = mean_mapping_quality, fill = social, colour = social)) +
   geom_density(alpha = 0.5) +
   labs(title = plot_title, subtitle = plot_subtitle) +
   # Monogynous orange f, Polygynous purple 9
   scale_fill_manual(values   = c("#f1a340", "#998ec3")) +
   scale_colour_manual(values = c("#f1a340", "#998ec3")) +
   theme_classic() +
   annotate("text", x = 21, y = 0.3, label = annotation_text)

## plot only samples from the same origin as reference
# Bruniquel and Perpignan
samples_to_plot <- subset(metadata_df, subset = population %in% c("Bruniquel", "Pyrenees"), select = sample.soc)$sample.soc
qualimap_stats_samples_in_ref_geographical <- subset(qualimap_stats, subset = sample_name %in% samples_to_plot)
# The null hypothesis is that both groups were sampled from populations with identical distributions
mean_mapping_qualityM <- qualimap_stats_samples_in_ref_geographical$mean_mapping_quality[qualimap_stats_samples_in_ref_geographical$social== "M"]
mean_mapping_qualityP <- qualimap_stats_samples_in_ref_geographical$mean_mapping_quality[qualimap_stats_samples_in_ref_geographical$social== "P"]
plot_title    <- paste("Figure 5: Mean mapping quality by social type")
plot_subtitle <- paste("Data from Qualimap report, samples from same region as reference")
annotation_text <- paste("KS test p =", round(ks.test(x = mean_mapping_qualityM, y = mean_mapping_qualityP)$p.value, digits = 5),
                         "\nWilcoxon test p =", round(wilcox.test(x = mean_mapping_qualityM, y = mean_mapping_qualityP)$p.value, digits = 5))

ggplot(qualimap_stats_samples_in_ref_geographical, aes(x = mean_mapping_quality, fill = social, colour = social)) +
   geom_density(alpha = 0.5) +
   labs(title = plot_title, subtitle = plot_subtitle) +
   # Monogynous orange f, Polygynous purple 9
   scale_fill_manual(values   = c("#f1a340", "#998ec3")) +
   scale_colour_manual(values = c("#f1a340", "#998ec3")) +
   theme_classic() +
   annotate("text", x = 21, y = 0.3, label = annotation_text)


```

The quality of mapping is statistically different between social group (KS test p = 0.002, Wilcoxon test p = 0.0003).
We hypothezise that the genome reference influences these results. We thus subset the data for the samples that share the geographical origins with the genome reference (Bruniquel and Pyrennes, 86 samples), the differences are statistically valid but to a lesser strength (KS test p = 0.04, Wilcoxon test p = 0.01). We conclude that the geographical origin has an impact on these results.



## Number of mismatches

I am checking if the number of mismatches are different between monogynous and polygynous alignments to the reference.

```{r number of mismatches, eval = TRUE, echo = FALSE, warning = FALSE}
## plot number of mismatches per social type
plot_title    <- paste("Figure 6: Number of mismatches per social type")
plot_subtitle <- paste("Data from Qualimap report")

# clean data
qualimap_stats$mismatches_number <- gsub(x = qualimap_stats$mismatches_number,
                                         pattern = ",",
                                         replacement = "")

qualimap_stats$mismatches_number <- as.numeric(gsub(x = qualimap_stats$mismatches_number,
                                         pattern = " ",
                                         replacement = ""))
# test for differences in means
print(paste("Number of mismatches is not significantly different between monogynous and polygynous samples, t.test p-value = ",
            t.test(x = qualimap_stats$mismatches_number[qualimap_stats$social == "M"],
                   y = qualimap_stats$mismatches_number[qualimap_stats$social == "P"])$p.value)) # 0.9800903

ggplot(data = qualimap_stats, aes(x = sample_name, y = mismatches_number, fill = social)) +
    geom_bar(colour = "black", stat = "identity") +
    guides(fill = FALSE) +
    labs(title = plot_title, subtitle = plot_subtitle) +
    # Monogynous orange f, Polygynous purple 9
    scale_fill_manual(values   = c("#f1a340", "#998ec3")) +
    scale_colour_manual(values = c("#f1a340", "#998ec3")) +
    theme_classic() + 
    theme(axis.text.x  = element_text(angle = 45, vjust = 0.5, size = 4.5))
```
## Conclusion

Polygynous samples differ significantly from monogynous samples, even before filtering: they have less raw sequences (Figure 1) and a lower mapping quality (Figure 4).
Next I will check the distributions after filtering.